src/BuiltInTools/dotnet-watch/HotReload/NamedPipeContract.cs

```diff
@@ -79,117 +63,369 @@ public static void WriteLog(Stream stream, IReadOnlyCollection<(string message,
-            if (version != Version)
+            var count = await stream.ReadInt32Async(cancellationToken);
+            if (count == 0)
             {
-                throw new NotSupportedException($"Unsupported version {version}.");
+                return [];
             }
 
-            var capabilities = binaryReader.ReadString();
-            return new ClientInitializationPayload(capabilities);
+            var bytes = new byte[count];
+            await ReadExactlyAsync(stream, bytes, cancellationToken);
+            return bytes;
+        }
+
+        public static async ValueTask<int[]> ReadIntArrayAsync(this Stream stream, CancellationToken cancellationToken)
+        {
+            var count = await stream.ReadInt32Async(cancellationToken);
+            if (count == 0)
+            {
+                return [];
+            }
+
+            var result = new int[count];
+            int size = count * sizeof(int);
+            var buffer = ArrayPool<byte>.Shared.Rent(minimumLength: size);
+            try
+            {
+                await ReadExactlyAsync(stream, buffer.AsMemory(0, size), cancellationToken);
+
+                for (var i = 0; i < count; i++)
+                {
+                    result[i] = BinaryPrimitives.ReadInt32LittleEndian(buffer.AsSpan(i * sizeof(int)));
+                }
+            }
+            finally
+            {
+                ArrayPool<byte>.Shared.Return(buffer);
+            }
+
+            return result;
+        }
+
+        public static async ValueTask<string> ReadStringAsync(this Stream stream, CancellationToken cancellationToken)
+        {
+            int size = await stream.Read7BitEncodedIntAsync(cancellationToken);
+            if (size < 0)
+            {
+                throw new InvalidDataException();
+            }
+
+            if (size == 0)
+            {
+                return string.Empty;
+            }
+
+            var buffer = ArrayPool<byte>.Shared.Rent(minimumLength: size);
+            try
+            {
+                await ReadExactlyAsync(stream, buffer.AsMemory(0, size), cancellationToken);
+                return Encoding.UTF8.GetString(buffer.AsSpan(0, size));
+            }
+            finally
+            {
+                ArrayPool<byte>.Shared.Return(buffer);
+            }
+        }
+
+        public static async ValueTask<int> Read7BitEncodedIntAsync(this Stream stream, CancellationToken cancellationToken)
+        {
+            const int MaxBytesWithoutOverflow = 4;
+
+            uint result = 0;
+            byte b;
+
+            for (int shift = 0; shift < MaxBytesWithoutOverflow * 7; shift += 7)
+            {
+                b = await stream.ReadByteAsync(cancellationToken);
+                result |= (b & 0x7Fu) << shift;
+
+                if (b <= 0x7Fu)
+                {
+                    return (int)result;
+                }
+            }
+
+            // Read the 5th byte. Since we already read 28 bits,
```

Comment:
Could this run across different versions, where the code herein has a different version between the "client" and the "server"?  If so, would there need to be tests with the serialized state, where that serialized state is checked in and effectively unchanging?